### DISCUSSION TITLE
How do I specify the Oracle Client library location (`lib_dir`) in SQLAlchemy's `create_engine`?

### DISCUSSION THREAD

**Discussion by ChemGuy88** [sqlalchemy/sqlalchemy#8323] [Usage Questions] [answered]
How do I specify the Oracle Client library location (`lib_dir`) in SQLAlchemy's `create_engine`?

According to [Passing cx_Oracle connect arguments](https://docs.sqlalchemy.org/en/14/dialects/oracle.html#passing-cx-oracle-connect-arguments), I should be able to pass cx_Oracle arguments via the `connect_args` dictionary to SQLAlchemy's `create_engine`, as such:

```python
engine = create_engine(ENGINE_PATH_WIN_AUTH,
                       connect_args={"lib_dir": lib_dir})
```

However, I still get the following error:

> DatabaseError: (cx_Oracle.DatabaseError) DPI-1047: Cannot locate a 64-bit Oracle Client library: "C:\Oracle\product\11.2.0\client\bin\oci.dll is not the correct architecture". See https://cx-oracle.readthedocs.io/en/latest/user_guide/installation.html for help
(Background on this error at: https://sqlalche.me/e/14/4xp6)

**Comment by zzzeek** [comment, marked as answer]
looks like it found the file you are passing and it's just not correct for some reason.  

you might want to use the new [oracledb thin client](https://pypi.org/project/oracledb/) instead.     check out https://levelup.gitconnected.com/using-python-oracledb-1-0-with-sqlalchemy-pandas-django-and-flask-5d84e910cb19

**Reply by ChemGuy88** [reply]
@zzzeek  @CaselIT The path it mentions, `C:\Oracle\product\11.2.0\client\bin`, is not the path I specific in `lib_dir`. And I know the client I specify in `lib_dir` works, because if I try to connect with `cx_Oracle.connect()`, it works.

```python
sys.platform  # 'win32'

import cx_Oracle
cx_Oracle.init_oracle_client(lib_dir=lib_dir)
dsn_tns = cx_Oracle.makedsn(ip, port, sid)
connection = cx_Oracle.connect(username, password, dsn_tns, encoding="UTF-8")
query = """SELECT * FROM myTable"""
cursor.execute(query)

for row in cursor:
    pass
```

**Reply by zzzeek** [reply]
SQLAlchemy doesn't call the `init_oracle_client()` method.  You should call that directly in your SQLAlchemy program if that's what's required.

on Linux we just use LD_LIBRARY_PATH for this kind of thing.

**Reply by ChemGuy88** [reply]
Okay thanks

**Reply by CaselIT** [reply]
Once v2 is out with oracledb you can actually make sqlalchemy call the init method, and you can also specify options for it:
```py
engine = create_engine('oracle+oracledb://...', thick_mode={"lib_dir": lib_dir})
```

**Reply by ChemGuy88** [reply]
Based on [this Google discussion](https://groups.google.com/g/comp.lang.tcl/c/J48G1yhvFrc), I added

```shell
set TCLLIBPATH="C:\path\to\instantclient_21_6"
```

to `user_settings.bat` in my anaconda virtual environment (`%CONDA_PREFIX%\etc\conda\activate.d`). However, I get the same error. The environment variable is being set, since

```python
import os
"TCLLIBPATH" in os.environ.keys()  # True
```

Do you have any more recommendations?

**Reply by CaselIT** [reply]
is adding a call to `cx_Oracle.init_oracle_client(lib_dir=lib_dir)` before creating the sqlalchemy engine not an option?

**Reply by ChemGuy88** [reply]
Oh, I didn't realize I could do that. It works now.

For posterity I did:

```python
cx_Oracle.init_oracle_client(lib_dir=lib_dir)

ENGINE_PATH_WIN_AUTH = ...  # Defined as in SQLAlchemy documentation
connection = create_engine(ENGINE_PATH_WIN_AUTH)
output = pd.read_sql_query(sql=query, con=connection)
```

Thanks!

**Reply by CaselIT** [reply]
Great!

**Comment by CaselIT** [comment]
Just guessing, but it seems it's looking for a 64bit version of the dialect and you may have installed the 32bit one.
Sadly 32 and 64bit are still a thing on windows and they are in general not compatible.
